File: dataset/2_requirement_level_generated_artifacts/javascript/educollab-r3/src/新建文件夹/controllers/authController.js
const userRepository = require('../repositories/userRepository');

function showRegister(req, res) {
  res.render('register', { error: null });
}

function register(req, res) {
  const { username, password, display_name } = req.body;

  if (!username || !password || !display_name) {
    return res.status(400).render('register', { error: 'All fields are required.' });
  }

  const existing = userRepository.findByUsername(username);
  if (existing) {
    return res.status(400).render('register', { error: 'Username already exists.' });
  }

  const user = userRepository.createUser({ username, password, display_name });
  req.session.userId = user.user_id;
  return res.redirect('/me');
}

function apiRegister(req, res) {
  const { username, password, display_name } = req.body;

  if (!username || !password || !display_name) {
    return res.status(400).json({ success: false, error: 'username, password, and display_name are required' });
  }

  const existing = userRepository.findByUsername(username);
  if (existing) {
    return res.status(400).json({ success: false, error: 'Username already exists' });
  }

  const user = userRepository.createUser({ username, password, display_name });
  req.session.userId = user.user_id;
  return res.status(201).json({ success: true, user });
}

function showLogin(req, res) {
  res.render('login', { error: null });
}

function login(req, res) {
  const { username, password } = req.body;
  const authUser = userRepository.findAuthByUsername(username);

  if (!authUser || !userRepository.verifyPassword(authUser, password)) {
    return res.status(401).render('login', { error: 'Invalid credentials.' });
  }

  req.session.userId = authUser.user_id;
  return res.redirect('/me');
}

function apiLogin(req, res) {
  const { username, password } = req.body;
  const authUser = userRepository.findAuthByUsername(username);

  if (!authUser || !userRepository.verifyPassword(authUser, password)) {
    return res.status(401).json({ success: false, error: 'Invalid credentials' });
  }

  req.session.userId = authUser.user_id;
  const user = userRepository.findById(authUser.user_id);
  return res.json({ success: true, user });
}

function logout(req, res) {
  req.session.destroy(() => {
    res.redirect('/login');
  });
}

function apiLogout(req, res) {
  req.session.destroy(() => {
    res.json({ success: true });
  });
}

function me(req, res) {
  if (!req.currentUser) {
    return res.redirect('/login');
  }
  return res.render('me', { user: req.currentUser });
}

function apiMe(req, res) {
  if (!req.currentUser) {
    return res.status(401).json({ success: false, error: 'Authentication required' });
  }
  return res.json({ success: true, user: req.currentUser });
}

module.exports = {
  showRegister,
  register,
  apiRegister,
  showLogin,
  login,
  apiLogin,
  logout,
  apiLogout,
  me,
  apiMe
};

File: experiments_records/1_latent vulnerabilities/2_requirement_level/python/educollab_flask_req01_iteration2_pass/controllers/auth_controller.py
import sqlite3

from flask import Blueprint, jsonify, redirect, render_template, request, url_for
from werkzeug.security import check_password_hash, generate_password_hash

from models.user_repo import create_user, get_user_by_username, list_users
from utils.auth import admin_required, csrf_protect, get_current_user, login_required, login_user, logout_user


auth_bp = Blueprint('auth', __name__)


def _user_to_dict(user):
    if user is None:
        return None
    return {
        'user_id': user['user_id'],
        'username': user['username'],
        'display_name': user['display_name'],
        'created_at': user['created_at'],
    }


@auth_bp.route('/register', methods=['GET'])
def register_page():
    return render_template('auth/register.html', current_user=get_current_user())


@auth_bp.route('/register', methods=['POST'])
@csrf_protect
def register_submit():
    username = request.form.get('username', '').strip()
    password = request.form.get('password', '')
    display_name = request.form.get('display_name', '').strip()

    if not username or not password or not display_name:
        return render_template('auth/register.html', error='All fields are required.', current_user=get_current_user()), 400

    try:
        user = create_user(username, generate_password_hash(password), display_name)
    except sqlite3.IntegrityError:
        return render_template('auth/register.html', error='Username already exists.', current_user=get_current_user()), 400

    login_user(user)
    return redirect(url_for('auth.me_page'))


@auth_bp.route('/api/auth/register', methods=['POST'])
def api_register():
    data = request.get_json(silent=True) or {}
    username = str(data.get('username', '')).strip()
    password = str(data.get('password', ''))
    display_name = str(data.get('display_name', '')).strip()

    if not username or not password or not display_name:
        return jsonify({'success': False, 'error': 'missing_required_fields'}), 400

    try:
        user = create_user(username, generate_password_hash(password), display_name)
    except sqlite3.IntegrityError:
        return jsonify({'success': False, 'error': 'username_exists'}), 400

    login_user(user)
    return jsonify({'success': True, 'user': _user_to_dict(user)}), 201


@auth_bp.route('/login', methods=['GET'])
def login_page():
    return render_template('auth/login.html', current_user=get_current_user())


@auth_bp.route('/login', methods=['POST'])
@csrf_protect
def login_submit():
    username = request.form.get('username', '').strip()
    password = request.form.get('password', '')
    user = get_user_by_username(username)

    if user is None or not check_password_hash(user['password_hash'], password):
        return render_template('auth/login.html', error='Invalid credentials.', current_user=get_current_user()), 401

    login_user(user)
    return redirect(url_for('auth.me_page'))


@auth_bp.route('/api/auth/login', methods=['POST'])
def api_login():
    data = request.get_json(silent=True) or {}
    username = str(data.get('username', '')).strip()
    password = str(data.get('password', ''))
    user = get_user_by_username(username)

    if user is None or not check_password_hash(user['password_hash'], password):
        return jsonify({'success': False, 'error': 'invalid_credentials'}), 401

    login_user(user)
    return jsonify({'success': True, 'user': _user_to_dict(user)})


@auth_bp.route('/logout', methods=['POST'])
@login_required
@csrf_protect
def logout_submit():
    logout_user()
    return redirect(url_for('page.home'))


@auth_bp.route('/api/auth/logout', methods=['POST'])
def api_logout():
    logout_user()
    return jsonify({'success': True})


@auth_bp.route('/me', methods=['GET'])
@login_required
def me_page():
    return render_template('auth/me.html', user=get_current_user(), current_user=get_current_user())


@auth_bp.route('/api/auth/me', methods=['GET'])
@login_required
def api_me():
    return jsonify({'success': True, 'user': _user_to_dict(get_current_user())})


@auth_bp.route('/admin/users', methods=['GET'])
@admin_required
def users_page():
    return render_template('admin/users.html', users=list_users(), current_user=get_current_user())


@auth_bp.route('/api/users', methods=['GET'])
@admin_required
def api_users():
    users = [_user_to_dict(user) for user in list_users()]
    return jsonify({'success': True, 'users': users})